### PR TITLE
#20216: Wiring should work in arbitrary nest levels far from original module

### LINKED ISSUE #20216: java.util.NoSuchElementException in GraphDSL
This happens in 2.4.3, but not in 2.4.2

```scala
import akka.actor.ActorSystem
import akka.stream.scaladsl.{Broadcast, Flow, GraphDSL, Sink, Source, Zip}
import akka.stream.{ActorMaterializer, ActorMaterializerSettings, FlowShape}

object AkkaTest extends App {

  implicit val as = ActorSystem("ReactiveKafka")
  implicit val materializer = ActorMaterializer(
    ActorMaterializerSettings(as)
      .withInputBuffer(16, 16)
      .withAutoFusing(false)
  )
  val subflow = GraphDSL.create() { implicit b =>
    import GraphDSL.Implicits._
    val zip = b.add(Zip[String, String]())
    val bc = b.add(Broadcast[String](2))

    bc.out(0) ~> zip.in0
    bc.out(1) ~> zip.in1

    FlowShape(bc.in, zip.out)
  }.named("subflow")

  val higherLevelFlow = Flow[String].via(subflow).named("higherLevelFlow")

  //fails
  Source(List("")).via(higherLevelFlow).to(Sink.ignore).run()

  //ok
  Source(List("")).via(subflow).to(Sink.ignore).run()
}
```

The flow which uses higherLevelFlow fails with this exception: 

```
Exception in thread "main" java.util.NoSuchElementException: key not found: ZipWith2.in0
	at scala.collection.MapLike$class.default(MapLike.scala:228)
	at scala.collection.AbstractMap.default(Map.scala:59)
	at scala.collection.MapLike$class.apply(MapLike.scala:141)
	at scala.collection.AbstractMap.apply(Map.scala:59)
	at akka.stream.impl.MaterializerSession.assignPort(StreamLayout.scala:911)
	at akka.stream.impl.MaterializerSession$$anonfun$akka$stream$impl$MaterializerSession$$exitScope$1.apply(StreamLayout.scala:803)
	at akka.stream.impl.MaterializerSession$$anonfun$akka$stream$impl$MaterializerSession$$exitScope$1.apply(StreamLayout.scala:802)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at akka.stream.impl.MaterializerSession.akka$stream$impl$MaterializerSession$$exitScope(StreamLayout.scala:802)
	at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:858)
	at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:850)
	at scala.collection.immutable.Set$Set2.foreach(Set.scala:128)
	at akka.stream.impl.MaterializerSession.materializeModule(StreamLayout.scala:850)
	at akka.stream.impl.MaterializerSession.materializeComposite(StreamLayout.scala:884)
	at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:860)
	at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:850)
	at scala.collection.immutable.Set$Set1.foreach(Set.scala:94)
	at akka.stream.impl.MaterializerSession.materializeModule(StreamLayout.scala:850)
	at akka.stream.impl.MaterializerSession.materializeComposite(StreamLayout.scala:884)
	at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:860)
	at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:850)
	at scala.collection.immutable.Set$Set1.foreach(Set.scala:94)
	at akka.stream.impl.MaterializerSession.materializeModule(StreamLayout.scala:850)
	at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:857)
	at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:850)
	at scala.collection.immutable.Set$Set3.foreach(Set.scala:163)
	at akka.stream.impl.MaterializerSession.materializeModule(StreamLayout.scala:850)
	at akka.stream.impl.MaterializerSession.materialize(StreamLayout.scala:817)
	at akka.stream.impl.ActorMaterializerImpl.materialize(ActorMaterializerImpl.scala:190)
	at akka.stream.impl.ActorMaterializerImpl.materialize(ActorMaterializerImpl.scala:79)
	at akka.stream.scaladsl.RunnableGraph.run(Flow.scala:367)
	at casino.kasha.AkkaTest$.delayedEndpoint$casino$kasha$AkkaTest$1(AkkaTest.scala:29)
	at casino.kasha.AkkaTest$delayedInit$body.apply(AkkaTest.scala:7)
	at scala.Function0$class.apply$mcV$sp(Function0.scala:34)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
	at scala.App$$anonfun$main$1.apply(App.scala:76)
	at scala.App$$anonfun$main$1.apply(App.scala:76)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35)
	at scala.App$class.main(App.scala:76)
	at casino.kasha.AkkaTest$.main(AkkaTest.scala:7)
	at casino.kasha.AkkaTest.main(AkkaTest.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
```

Flow which uses `subflow` directly works fine.

### PR DESCRIPTION
**Pull request by drewhk**
Fixes #20216

### CODE REVIEW

**Comment by patriknw** [comment]
LGTM

**Comment by drewhk** [comment]
I am not sure now about this one. It fixes the issue but I am not sure now if it is 100% correct.

**Comment by drewhk** [comment]
Ok, I know why now.

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala [hunk 1]
```diff
@@ -906,19 +906,25 @@ private[stream] abstract class MaterializerSession(val topLevel: StreamLayout.Mo
 
   final protected def assignPort(in: InPort, subscriberOrVirtual: AnyRef): Unit = {
     subscribers.put(in, subscriberOrVirtual)
-    // Interface (unconnected) ports of the current scope will be wired when exiting the scope
-    if (!currentLayout.inPorts(in)) {
-      val publisher = publishers.get(currentLayout.upstreams(in))
-      if (publisher ne null) doSubscribe(publisher, subscriberOrVirtual)
+
+    currentLayout.upstreams.get(in) match {
```
Comment (2beaucoup) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: `foreach`?

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: doesn't that allocate a closure?

Comment (2beaucoup) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: I guess yes. But it looks moar nice. :)

Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: you can't have it all :)

**Pull request merged**